### PR TITLE
docs: add messaging_reference.adoc

### LINKED ISSUE #1086: Request for Contribution: L1-L2 Messaging programming reference
_Please reply in the comments if you are interested._

# Description

Document the L1 functions and L1 events that enable L1->L2 messaging. Include general descriptions of the contracts that include the functions. For example, see the [StarkGate function and event reference](https://docs.starknet.io/documentation/tools/starkgate_function_reference/)

## For each _function_, include the following information:

- `<function name>`
- **Description**: An in-depth description of the function that includes:
  - What does the function do?
  - Why does it matter?
  - Any other information that is necessary or useful to know.
- **State mutability**: `Pure`/`payable`/`non-payable`/`view`
- **Parameters**: For each parameter: <type> <param name>:: <Description>. If there are no parameters, specify None.
- **Return values**: Same as Parameters. If there are no return values, specify None.
- **Function definitions**: Links to the function interfaces within the contract, and to the function implementations.

Examples: [`deposit`](https://docs.starknet.io/documentation/tools/starkgate_function_reference/#deposit),  [`getStatus`](https://docs.starknet.io/documentation/tools/starkgate_function_reference/#getStatus), and others in the _StarkGate function and event reference_.

## For each _event_, include the following information:

- `<event name>`
- **Description**: 
  - What does the function do?
  - Why does it matter?
  - Any other information that is necessary or useful to know.
- **Event attributes**: List attributes of the event.

Examples: See **Emitted event** and **Event attributes** in [`depositCancelRequest`](https://docs.starknet.io/documentation/tools/starkgate_function_reference/#depositCancelRequest), [`withdraw`](https://docs.starknet.io/documentation/tools/starkgate_function_reference/#withdraw), and others  in the _StarkGate function and event reference_.

**NOTE:** In these examples, the events are directly tied to the relevant functions. For the Messaging event reference, the events should have their own dedicated section.

## Source material

- You can find the interfaces for these functions in [IStarknetMessaging.sol](https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessaging.sol)
- L1 Events: [IStarknetMessagingEvents.sol](https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessagingEvents.sol)
- L1 Function implementations (For the implementation, you just need to provide a link to the implementation, nothing more): [StarknetMessaging.sol](https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol).

# Instructions for Contributing

1. Leave a comment on this Issue indicating that you want to take ownership of this feature.
2. Wait for our confirmation to start on the project.
3. Once you have received confirmation, begin working on your contribution in a new branch.
4. Create a new AsciiDoc file named `messaging_reference.adoc` in `components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture`. Use the code in the Template code section. Stick to this format.
5. Add the new file to the table of contents by adding an `xref` to it in `components/Starknet/modules/architecture_and_concepts/nav.adoc`
6. Follow these [basic writing guidelines](https://github.com/starknet-io/starknet-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#basic-writing-guidelines).
6. Be ready to make changes based on review comments.

## AsciiDoc Function Template

```
[id=”messaging_reference”]
= Messaging reference

== L1 Function reference
'''

[#function_name]
=== `<function_name>`

[discrete]
==== Description

<Add text for description here>

[discrete]
==== State Mutability

`payable` | `non-payable` | `view` | `pure`

[discrete]
==== Parameters

[horizontal,labelwidth="30",role=stripes-odd]
`<type> _<param_name>_`:: <Description>

[discrete]
==== Returns

This section should include a list of return values and what they mean in the same format as Parameters:

[horizontal,labelwidth="30",role=stripes-odd]
`<type> _<return_value>_`:: <Description>


[discrete]
==== Function definitions

Contract: link:<URL>[`<filename>.sol`]

* Function: link:<URL_permalink>[`<function_name>`]

```

## AsciiDoc Event Template

Add this below all functions

```
== L1 Event reference

Add link to contract here.

[#event_name]
=== `<event_name>`

[discrete]
==== Description

<Add text for description here, including what triggers this event.>

[discrete]
==== Event attributes

[horizontal,role=stripes-odd]
`<type> _<attribute_name>_`:: <Description>.

[discrete]
==== Event definition

* Contract: link:<URL_of_contract>[`<contract_name>`]

* Event: link:<permalink_to_event>[`<event_name>`]
```

# Remuneration

Compensation for the approved contribution will be provided via OnlyDust.

# Deadline

Maximum of 10 days per example

# Contact

If you have any questions or need further details, please contact @stoobie.

__________________
We look forward to your proposals. Thank you for contributing to the Starknet ecosystem!

### PR DESCRIPTION
**Pull request by xiaolou86**
### Description of the Changes

Add messaging_reference.adoc

Closes #1086 

### PR Preview URL

[Messaging reference](https://starknet-io.github.io/starknet-docs/pr-1094/documentation/architecture_and_concepts/Network_Architecture/messaging_reference/)

### Check List

- [x] Changes have been done against main branch, and PR does not conflict
- [x] PR title follows the convention: `<docs/feat/fix/chore>(optional scope): <description>`, e.g: `fix: minor typos in code`

- - -
This change is [Reviewable](https://reviewable.io/reviews/starknet-io/starknet-docs/1094)

### CODE REVIEW

**Comment by xiaolou86** [comment]
@stoobie I finished the doc, please help me to review. Thank you very much.

Failed to generate a preview, it's a little unconvinient.

**Comment by stoobie** [comment]
> Failed to generate a preview, it's a little inconvenient.

I know. Are you able to generate a preview locally?

**Comment by xiaolou86** [comment]
> > Failed to generate a preview, it's a little inconvenient.
> 
> I know. Are you able to generate a preview locally?

I tried in my other repo, but it failed. I will try in my forked starknet-docs repo later.

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+A message may only be canceled by its sender.
+
+If the message is missing, the call will revert.
+
+Note that the message fee is not refunded.
+
+[discrete]
+==== State Mutability
+
+None.
```
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:70]: I need to find out the state mutability of these functions.

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+
+[discrete]
+==== Parameters
+
+[horizontal,labelwidth="30",role=stripes-odd]
+`uint256 _fromAddress_`:: The address of the L2 contract.
```
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:115]:
```suggestion
`uint256 _fromAddress_`:: The address of the L2 contract that sent the message.
```

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+
+'''
+
+[#getMaxL1MsgFee]
+=== `getMaxL1MsgFee`
+
+[discrete]
+==== Description
+
+Returns the max fee (in Wei) that StarkNet will accept per single message.
```
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:139]:
```suggestion
Returns the maximum fee, in Wei, that Starknet accepts per single message.
```

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+[discrete]
+==== Parameters
+
+None.
+
+[discrete]
+==== Returns
+
+[horizontal,labelwidth="30",role=stripes-odd]
+`uint256 _maxFee_`:: The max fee (in Wei) that StarkNet will accept per single message.
```
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:155]:
```suggestion
`uint256 _maxFee_`:: The maximum fee, in Wei, that Starknet accepts per single message.
```

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+'''
+
+[#l1ToL2Messages]
+=== `l1ToL2Messages`
+
+[discrete]
+==== Description
+
+Returns `msg_fee + 1` if there is a pending message associated with the given 'msgHash', otherwise, returns 0.
```
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:172]:
```suggestion
[horizontal,labelwidth="30",role=stripes-odd]
`msg_fee + 1`:: A pending message is associated with the given message hash.
`0`:: No pending message is associated with the given message hash.
```

Comment (stoobie) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:172]: I'll continue reviewing from here.

**Review by stoobie — CHANGES_REQUESTED** [review]
Began review. Will continue.

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 6]
```diff
@@ -0,0 +1,444 @@
+[id="messaging_reference"]
+= Messaging reference
+:description: Comprehensive function and event reference for StarknetMessaging.
+
+The StarknetMessaging smart contracts include functions and events that you use to implement various flows in a dApp.
+
+For information on the messaging mechanism, see xref:messaging-mechanism.adoc[Messaging mechanism].
```
[the remark this answers is not in the document]
Comment (xiaolou86) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:7]: Thanks for your guide.

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+=== `cancelL1ToL2Message`
+
+[discrete]
+==== Description
+
+Cancels an L1 to L2 message. This function should be called at least messageCancellationDelay() seconds after the call to startL1ToL2MessageCancellation().
+
+A message may only be canceled by its sender.
+
+If the message is missing, the call will revert.
```
[the remark this answers is not in the document]
Comment (xiaolou86) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:63]: the message is not exactly the `payload` parameter, it's a hash value.

**Comment by stoobie** [comment]
@xiaolou86 I apologize for not reviewing this more quickly. I'm getting back to this now.

**Comment by stoobie** [comment]
@xiaolou86 Would you please rebase your branch to merge all recent commits to starknet-io:main and commit again?

I want to test your fix to generate a preview from a fork.

**Comment by xiaolou86** [comment]
> @xiaolou86 Would you please rebase your branch to merge all recent commits to starknet-io:main and commit again?
> 
> I want to test your fix to generate a preview from a fork.

`Preview generated` successfully, while `Send comment to PR` didn't work well :(

**Comment by xiaolou86** [comment]
I think maybe need to add `pull-requests: write` permission in the .yml file

**Comment by stoobie** [comment]
@xiaolou86 This has taken me a lot longer than I expected to move forward. To be fair to you, I sent you initial compensation just now for the work you've done so far on this, and I'll send more once we finish this.

**Comment by xiaolou86** [comment]
@stoobie Thank you very much.

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 8]
```diff
@@ -0,0 +1,446 @@
+[id="messaging_reference"]
+= Messaging reference
+:description: Comprehensive function and event reference for Starknet messaging.
+
+The Starknet messaging smart contracts include functions and events that you use to implement various flows in a dApp.
```
Comment (LucasLvy) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:5]: this needs to be rewritten to match the rest of the docs style

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+|https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/StarknetMessaging.sol[`StarknetMessaging.sol`] |
+The contract that implements messaging functions.
+
+Interfaces are available through https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/solidity/IStarknetMessaging.sol[`IStarknetMessaging.sol`]. a|
+* xref:#cancelL1ToL2Message[`cancelL1ToL2Message`]
+* xref:#consumeMessageFromL2[`consumeMessageFromL2`]
+* xref:#getMaxL1MsgFee[`getMaxL1MsgFee`]
+* xref:#l1ToL2Messages[`l1ToL2Messages`]
+* xref:#sendMessageToL2[`sendMessageToL2`]
+* xref:#startL1ToL2MessageCancellation[`startL1ToL2MessageCancellation`]
```
Comment (LucasLvy) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:24]: @stoobie sorted alphabetically not like the contract definition

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+
+'''
+
+[#cancelL1ToL2Message]
+=== `cancelL1ToL2Message`
+
+[discrete]
+==== Description
+
+Cancels an L1 to L2 message. Call this function after calling the `startL1ToL2MessageCancellation()` function. The time between the calls to these two functions must be at least the number of seconds defined by the link:https://github.com/starkware-libs/cairo-lang/blob/caba294d82eeeccc3d86a158adb8ba209bf2d8fc/src/starkware/starknet/solidity/StarknetMessaging.sol#L25[`L1L2_MESSAGE_CANCELLATION_DELAY_TAG`] constant.
```
Comment (LucasLvy) [inline components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:59]: Sort of correct, the constant defines the storage slot the value is actually stored at. To get the real value one should call [this](https://github.com/starkware-libs/cairo-lang/blob/caba294d82eeeccc3d86a158adb8ba209bf2d8fc/src/starkware/starknet/solidity/StarknetMessaging.sol#L59-L61) function

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc (hunk 9 above)
Comment (stoobie) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:24]: Are you requesting that they be sorted as shown in the contract?

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc (hunk 10 above)
Comment (stoobie) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:59]: So should we add `messageCancellationDelay()` to the reference?

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc (hunk 9 above)
Comment (LucasLvy) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc:24]: No I don’t care, just not sure how it’s done elsewhere

**Comment by xiaolou86** [comment]
@stoobie grateful for you helping me create a new PR #1209 to edit this messaging_reference.adoc? It's my duty. Thank you very much!

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc (hunk 10 above)
Comment (xiaolou86) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc]: I have added `messageCancellationDelay`

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc (hunk 9 above)
Comment (xiaolou86) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc]: In other docs, for example starkgate_function_reference.adoc, functions are listed in alphabetical order.

**Comment by xiaolou86** [comment]
@stoobie @LucasLvy  thank you for your reviews. 

I have updated the docs:
1. add the functions implemented in StarknetMessaging.sol but not part of the interface: 
```
l1ToL2MessageCancellations
l1ToL2MessageNonce
l2ToL1Messages
messageCancellationDelay
```
2. Adding a link to the `messageCancellationDelay` function on etherscan to get the real value.

File: components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc (hunk 8 above)
Comment (stoobie) [reply components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging_reference.adoc]: Responded in #1209.

**Comment by stoobie** [comment]
@xiaolou86 Please do the following:
1. Rebase from `starknet-io/starknet-docs:main`. I made a lot of changes to the directory structure and file names, including change the name of this file to `messaging-reference.adoc` (changed the underscore to a hyphen in the file name.)
2. Copy the changes over from #1209.
3. For each event that is emitted, state which function emits it. For each function that emits an event, add a heading "Emitted event" with the name of the event that emitted it, with a link to the event lower down in the page. For example, see `ConsumedMessageToL2` and `consumeMessageFromL2` in #1209.

**Comment by xiaolou86** [comment]
> @xiaolou86 Please do the following:
> 
> 1. Rebase from `starknet-io/starknet-docs:main`. I made a lot of changes to the directory structure and file names, including change the name of this file to `messaging-reference.adoc` (changed the underscore to a hyphen in the file name.)
> 2. Copy the changes over from [feat: Starknet messaging function and event reference (based on #1094) #1209](https://github.com/starknet-io/starknet-docs/pull/1209).
> 3. For each event that is emitted, state which function emits it. For each function that emits an event, add a heading "Emitted event" with the name of the event that emitted it, with a link to the event lower down in the page. For example, see `ConsumedMessageToL2` and `consumeMessageFromL2` in [feat: Starknet messaging function and event reference (based on #1094) #1209](https://github.com/starknet-io/starknet-docs/pull/1209).

@stoobie All done. Thank you very much.

**Comment by stoobie** [comment]
This has been reviewed by Lucas and Remo. I believe that I have implemented all their comments. So I am merging this.

**Review by stoobie — APPROVED** [review]
LGTM.

**Comment by stoobie** [comment]
@xiaolou86 Thank you for the contribution! I sent you a reward on Only Dust.

**Comment by xiaolou86** [comment]
@stoobie Thank you very much.

**Pull request merged**